### PR TITLE
Enabling support of OpenThread on TI CC1352P7

### PR DESCRIPTION
**Pull request by anobli**
This adds support of OpenThread to TI CC1352 and updates the OpenThread coprocessor application to support it the TI CC1352P7 Launchpad.
I used the CC1352P7 launchpad because I had a few of them in my closet).
I have only tested OpenThread on the launchpad (and on the beagleplay) but it should work on with all CC13x2 and CC26x2 SoC.

### CODE REVIEW

File: drivers/flash/soc_flash_cc13xx_cc26xx.c
Comment (cfriedt) [inline drivers/flash/soc_flash_cc13xx_cc26xx.c:187]: Hmm... It's fairly uncommon to dynamically allocate in drivers in Zephyr.

It might make sense to create an array of some fixed size in the `struct data` driver object, since that will always live in ram, and then initialize that location in the driver `init()` with contents that would be in devicetree / rom.

**Review by cfriedt — CHANGES_REQUESTED** [review]
Please drop usage of `k_malloc()` / `k_free()` for fixed-size buffers.

**Review by cfriedt — COMMENTED** [review]
Can probably drop the malloc Kconfig option, but other than that, LGTM.

For some reason, I was kind of surprised that we had never run OpenThread on this platform.

Really useful though - merçi 👍

File: drivers/flash/Kconfig.cc13xx_cc26xx
Comment (cfriedt) [inline drivers/flash/Kconfig.cc13xx_cc26xx:16]: Coding guideline checks are complaining about this
```suggestion
	bool "Flash to flash write"
```

File: samples/subsys/settings/boards/cc1352p7_lp.conf
Comment (fgrandel) [inline samples/subsys/settings/boards/cc1352p7_lp.conf:1]: Is this really a good single board default? I'd rather propose we make this the default for all TI SoCs that support flash/nvs (probably all of them CC @vaishnavachath), i.e. rename the file accordingly.

File: samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay
Comment (fgrandel) [inline samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay:1]: Can you please check which other TI SoC's have support for Thread now and rename the overlay accordingly? This should be all 13xx/26xx SoCs if I'm not mistaken (CC @vaishnavachath). (Don't forget to change the commit message, too.)

File: samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay
Comment (fgrandel) [inline samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay:1]: Optional: You may consider moving this configuration except for the "status = okay" to the board configuration of all supported cc13xx_cc16xx boards (with proper default PIN assignemnts), so that they all support the sample (and applications derived from it) mostly ootb and then only keep the "status=okay" here. @vaishnavachath Do you see any problems with this?

**Review by ghost — CHANGES_REQUESTED** [review]
Hi @anobli !

Thanks for this PR. It is so cool to see CC13/26xx boards support Thread. 8-) As IEEE 802.15.4 subsystem maintainer I'd love to promote your work quite a bit for others in the community to see how (relatively) simple it is to add basic Thread 1.1 support to existing drivers. Would you be willing to participate in one of Benjamin's tech talks on this topic together with me?

If you consider going even further to support Thread 1.2+ (which requires timed TX/RX) let me know. Robert and I invested considerable time and effort to ensure that current versions of Thread can be supported rather easily on additional platforms. Plus I've done some preliminary work on precision timing and timed TX/RX with the CC13/26xx platform. So it wouldn't be too much work plus I'll help contribute if required.

Guess it would make quite a splash in the water if suddenly a second platform supports the full Thread stack in Zephyr. :-)

**Comment by fgrandel** [comment]
CC @vaishnavachath @rlubos

**Comment by fgrandel** [comment]
@rlubos @jukkar Any ideas why that HTTP test fails?

File: drivers/flash/soc_flash_cc13xx_cc26xx.c
Comment (anobli) [reply drivers/flash/soc_flash_cc13xx_cc26xx.c:188]: My main concern was the memory usage. Except that, I have no reason to keep the switch.

File: samples/subsys/settings/boards/cc1352p7_lp.conf
Comment (anobli) [reply samples/subsys/settings/boards/cc1352p7_lp.conf:1]: Do you mean renaming and moving the file to socs ?
Something like socs/cc1352p7.conf (and the same for the other variant) ?

File: samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay
Comment (anobli) [reply samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay:1]: As far I know, if I rename the file to something like cc13xx_cc26xx.overlay, it won't be loaded automatically by zephyr build system so we will have to use DTC_OVERLAY_FILE to load it. Is it what you want ?

File: samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay
Comment (anobli) [reply samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay:1]: I moved everything related to uart1 to boards dts. Still, I am keeping the chosen node because I swap uart1 and uart0, which in my opinion make the application more user friendly.

**Comment by anobli** [comment]
> Thanks for this PR. It is so cool to see CC13/26xx boards support Thread. 8-) As IEEE 802.15.4 subsystem maintainer I'd love to promote your work quite a bit for others in the community to see how (relatively) simple it is to add basic Thread 1.1 support to existing drivers. Would you be willing to participate in one of Benjamin's tech talks on this topic together with me?

Hello, 
Sorry for the delay, I was offline.
Yes, I would be happy, if this not too late to participate to a tech talk with you.

> If you consider going even further to support Thread 1.2+ (which requires timed TX/RX) let me know. Robert and I invested considerable time and effort to ensure that current versions of Thread can be supported rather easily on additional platforms. Plus I've done some preliminary work on precision timing and timed TX/RX with the CC13/26xx platform. So it wouldn't be too much work plus I'll help contribute if required.

Yes, that's something I am considering. I am still learning how Thread and IEEE 802.15.4 are working so I won't work on it before a couple of week.

File: drivers/flash/soc_flash_cc13xx_cc26xx.c
Comment (fgrandel) [reply drivers/flash/soc_flash_cc13xx_cc26xx.c]: I see your point. Could you make the write buffer a local variable and encapsulate the whole logical branch in its own function? Then the memory would be taken from the stack if (and only if) the feature is actually being used. And probably the compiler finds enough optimization potential then to do what is right depending on the application.

File: samples/subsys/settings/boards/cc1352p7_lp.conf
Comment (fgrandel) [reply samples/subsys/settings/boards/cc1352p7_lp.conf]: Yes, there should now be a way to define [per-soc overlays](https://docs.zephyrproject.org/latest/build/kconfig/setting.html#the-initial-configuration) (in your case in `samples/subsys/settings/soc`):

> Furthermore, applications can have SoC overlay configuration that is applied to it, the file `socs/<SOC>_<BOARD_QUALIFIERS>.conf` will be applied if it exists, after the main project configuration has been applied and before any board overlay configuration files have been applied.

I've not yet tried this myself, but I guess it just works.

File: samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay
Comment (fgrandel) [reply samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay:1]: Ok, sounds good. I'll re-check once this is ready for final review.

**Comment by fgrandel** [comment]
I added a few more ideas in response to your questions/comments. Let me know when I should re-review the whole thing. I also invited another review from @cfriedt because I think you already fixed his concerns.

> Hello, Sorry for the delay, I was offline. Yes, I would be happy, if this not too late to participate to a tech talk with you.

It's not too late. Take your time. Once we have this PR merged we can think of how we want to present it.

Are you already using this locally in some Thread setup successfully?

>> If you consider going even further to support Thread 1.2+

> Yes, that's something I am considering. I am still learning how Thread and IEEE 802.15.4 are working so I won't work on it before a couple of week.

If I can somehow help you to get up to speed, let me know.

**Comment by fgrandel** [comment]
fyi: just saw that Chris is on holidays until the 24th, I'm sure he'll re-review as soon as he's back...

File: samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay
Comment (anobli) [reply samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay:1]: I thought a lot about it and I think that moving pinctrl and uart1 definitions to board folder should be enough.
Once this is done, everything remaining is related to the board config (although this will be probably be common for many development board).

File: drivers/flash/soc_flash_cc13xx_cc26xx.c
Comment (anobli) [reply drivers/flash/soc_flash_cc13xx_cc26xx.c]: Actually, I think it make sense.
I will also decrease the size of the buffer: 32 bytes should e enough.

**Comment by anobli** [comment]
> Are you already using this locally in some Thread setup successfully?

Yes! This is still WIP in progress but I use a beagleplay as OTBR.
I am also using a beagleconnect and a cc1352p7 lp as FTD.
I made an application layer that allows me using a device in homeassistant:
https://github.com/anobli/zephyr_esphome.git
This works (I can control LEDS and get temperature) but this is not reliable yet.

File: samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay
Comment (fgrandel) [reply samples/net/openthread/coprocessor/boards/cc1352p7_lp.overlay:1]: That's fine.

**Review by cfriedt — DISMISSED** [review]
LGTM - the only suggestion I have is to consider a for loop for the flash buffered write, but it's not blocking.

**Comment by fgrandel** [comment]
@anobli Is this ready for final review? If so please click the "re-request review" button in the upper right corner. It would be nice if we could merge this change.

**Comment by anobli** [comment]
> @anobli Is this ready for final review? If so please click the "re-request review" button in the upper right corner. It would be nice if we could merge this change.

Yes, but I would like to retest everything to be sure.
I have been able to build a RCP firmware for my beagleplay without issues but it has been more complicated for the cc1352p7_lp.
I am doing this on my spare time, and I have forgotten what I was doing for testing it ...
I hope finishing testing it tomorrow.

**Comment by fgrandel** [comment]
> I would like to retest everything to be sure.

No problem at all - take all the time you need. I just wanted to make sure you're not waiting for something from me.

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx.h
Comment (fgrandel) [inline drivers/ieee802154/ieee802154_cc13xx_cc26xx.h:83]: Not all chips covered by this driver actually have a PA. Of course this was a bug before but I think now it's a bug in reverse. See the solution in the cc13xx_26xx SubG driver which you could apply here as well.

Note: I'm really sorry because I'm quite sure my initial review on that type was not helpful either. :-( I didn't see that there were different types required because those lines did not appear in the diff. So you'll also have to re-introduce the #ifdef I asked you to remove, because those properties are probably missing.

Here is my original comment with some context:

> ```
> #if defined(CONFIG_SOC_CC1352P) || defined(CONFIG_SOC_CC1352P7)
> 		.pRegOverrideTxStd = NULL,
> ```
> This code seems to be redundant. As a statically initialized struct, it will be zeroed anyway. Switching to the correct structure as you've done in the .h file should be enough to fix the bug, right?

**Review by ghost — COMMENTED** [review]
Just minor remaining problems but probably still worth a fix - one of those had been part of a prior review the other one is my own fault I'm afraid.

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx.h
Comment (anobli) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx.h:83]: > See the solution in the cc13xx_26xx SubG driver which you could apply here as well.

Should I also copy rf_prop_overrides_tx_std and rf_prop_overrides_tx_20 ?
Are their values also valid for 2.4G radio ?

Comment (fgrandel) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx.h:83]: > Should I also copy rf_prop_overrides_tx_std and rf_prop_overrides_tx_20 ? Are their values also valid for 2.4G radio ?

No, I meant "similarly to", not copy over verbatim. But at the same time I was causing confusion because I made an error myself:

> So you'll also have to re-introduce the #ifdef I asked you to remove, because those properties are probably missing.
> ```
> #if defined(CONFIG_SOC_CC1352P) || defined(CONFIG_SOC_CC1352P7)
> 		.pRegOverrideTxStd = NULL,
> ```

You can still leave that one out. It is still initialized to NULL and therefore redundant. I have no idea what made me say that it has to be re-introduced other than I must have been really tired. Sigh.

What you have to change, though is something like this:

```diff
-volatile rfc_CMD_RADIO_SETUP_t cmd_radio_setup;
+#if defined(CONFIG_SOC_CC1352R) || defined(CONFIG_SOC_CC2652R) || defined(CONFIG_SOC_CC1352R7) || defined(CONFIG_SOC_CC2652R7)
+volatile rfc_CMD_RADIO_SETUP_t cmd_radio_setup;
+#elif defined(CONFIG_SOC_CC1352P) || defined(CONFIG_SOC_CC2652P) || defined(CONFIG_SOC_CC1352P7) || defined(CONFIG_SOC_CC2652P7)
+volatile rfc_CMD_RADIO_SETUP_PA_t cmd_radio_setup;
+#else
+BUILD_ASSERT(false, "unknown model");
+#endif /* CONFIG_SOC_CCxx52x */
```

And similarly in `ieee802154_cc13xx_cc26xx.c` around line 780:

```diff
-		.commandNo = CMD_RADIO_SETUP,
+#if defined(CONFIG_SOC_CC1352R) || defined(CONFIG_SOC_CC2652R) || defined(CONFIG_SOC_CC1352R7) || defined(CONFIG_SOC_CC2652R7)
+		.commandNo = CMD_RADIO_SETUP,
+#elif defined(CONFIG_SOC_CC1352P) || defined(CONFIG_SOC_CC2652P) || defined(CONFIG_SOC_CC1352P7) || defined(CONFIG_SOC_CC2652P7)
+		.commandNo = CMD_RADIO_SETUP_PA,
+#endif /* CONFIG_SOC_CCxx52x */
```

That is what I meant is similar to what we do in the SubG driver. But the SubG driver obviously doesn't include the 26xx models because they only have a 2.4G radio. And it uses slightly different commands.

Makes sense now?

**Review by ghost — APPROVED** [review]
Thanks  for your contribution and your patience. :-)

**Comment by fgrandel** [comment]
@anobli The test failures you're seeing seem to be unrelated. I mentioned them in the #stepper channel on Discord. Let's hope that they'll fix it soon and you can then rebase.

**Review by cfriedt — APPROVED** [review]

**Comment by cfriedt** [comment]
Looks like everything is passing

**Review by vaishnavachath — APPROVED** [review]

**Comment by fgrandel** [comment]
@anobli Still motivated to do a tech talk on this innovation? I've no other way to contact you - but if you're interested you can answer here or contact me on Discord (fgrandel).

**Pull request merged**
